test/lit/passes/cfp-rmw.wast

```diff
@@ -0,0 +1,898 @@
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $0)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (block (result i32)
+  ;; CHECK-NEXT:    (block (result i32)
+  ;; CHECK-NEXT:     (drop
+  ;; CHECK-NEXT:      (ref.as_non_null
+  ;; CHECK-NEXT:       (local.get $1)
+  ;; CHECK-NEXT:      )
+  ;; CHECK-NEXT:     )
+  ;; CHECK-NEXT:     (atomic.fence)
+  ;; CHECK-NEXT:     (i32.const 0)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (atomic.fence)
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $rmw-xchg-copy-fallthrough (param (ref $A) (ref $A)) (result i32)
+    ;; Same, and it works even with fallthrough.
+    (struct.atomic.rmw.xchg $A 0
+      (local.get 0)
+      (block (result i32)
+        (struct.atomic.get $A 0
+          (local.get 1)
+        )
+      )
+    )
+  )
+
+  ;; CHECK:      (func $rmw-xchg-copy-acqrel (type $1) (param $0 (ref $A)) (param $1 (ref $A)) (result i32)
+  ;; CHECK-NEXT:  (struct.atomic.rmw.xchg acqrel acqrel $A 0
+  ;; CHECK-NEXT:   (local.get $0)
+  ;; CHECK-NEXT:   (struct.atomic.get acqrel $A 0
+  ;; CHECK-NEXT:    (local.get $1)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $rmw-xchg-copy-acqrel (param (ref $A) (ref $A)) (result i32)
+    ;; Making the accesses acqrel instead of seqcst means that the replacement
+    ;; fence could be more expensive than the original op, so we don't optimize.
+    (struct.atomic.rmw.xchg acqrel acqrel $A 0
+      (local.get 0)
+      (struct.atomic.get acqrel $A 0
+        (local.get 1)
+      )
+    )
+  )
+
+  ;; CHECK:      (func $get (type $3) (param $0 (ref $A)) (result i32)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (ref.as_non_null
+  ;; CHECK-NEXT:    (local.get $0)
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (i32.const 0)
+  ;; CHECK-NEXT: )
+  (func $get (param (ref $A)) (result i32)
+    ;; This get is optimizable.
+    (struct.get $A 0
+      (local.get 0)
+    )
+  )
+)
+
+;; Not all rmw.xchg instructions are optimizable, though.
+(module
+  ;; CHECK:      (type $A (shared (struct (field (mut i32)))))
+  (type $A (shared (struct (mut i32))))
+
+  ;; CHECK:      (type $1 (func (param (ref $A)) (result i32)))
+
+  ;; CHECK:      (type $2 (func))
+
+  ;; CHECK:      (func $init (type $2)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.new_default $A)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $init
+    (drop
+      (struct.new_default $A)
+    )
+  )
+
+  ;; CHECK:      (func $rmw-xchg-mutate (type $1) (param $0 (ref $A)) (result i32)
+  ;; CHECK-NEXT:  (struct.atomic.rmw.xchg $A 0
+  ;; CHECK-NEXT:   (local.get $0)
+  ;; CHECK-NEXT:   (i32.const 1)
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $rmw-xchg-mutate (param (ref $A)) (result i32)
+    ;; The xchg mutates $A, so cannot be optimized out.
```

Comment:
Would it be more precise to say that it writes a second value (1, vs the default 0 that was there initially), so it isn't a constant field?